monix-eval/shared/src/main/scala/monix/eval/Task.scala

```diff
@@ -570,6 +570,112 @@ sealed abstract class Task[+A] extends Serializable {
+    * that is already closed due to the task being cancelled, thus triggering
+    * an error in the background with nowhere to go but in
+    * [[monix.execution.Scheduler.reportFailure Scheduler.reportFailure]].
+    *
+    * In this particular example, given that we are reading from a file,
+    * it doesn't matter. But in other cases it might matter, as concurrency
+    * on top of the JVM when dealing with I/O might lead to corrupted data.
+    *
+    * For those cases you might want to do synchronization (e.g. usage of
+    * locks and semaphores) and you might want to use [[bracketE]], the
+    * version that allows you to differentiate between nuormal termination
+    * and cancellation.
+    *
+    * @param use is a function that evaluates the resource yielded by the source,
+    *        yielding a result that will get generated by this function on
+    *        evaluation
+    *
+    * @param release is a function that gets called after `use` terminates,
+    *        either normally or in error, or if it gets cancelled, receiving
+    *        as input the resource that needs that needs release
+    */
+  final def bracket[B](use: A => Task[B])(release: A => Task[Unit]): Task[B] =
+    bracketE(use)((a, _) => release(a))
+
+  /** Returns a task that treats the source task as the acquisition of a resource,
+    * which is then exploited by the `use` function and then `released`, with
+    * the possibility of distinguishing between normal termination and cancellation,
+    * such that an appropriate release of resources can be executed.
+    *
+    * The `bracketE` operation is the equivalent of `try {} catch {} finally {}`
+    * statements from mainstream languages.
+    *
+    * The `bracketE` operation installs the necessary exception handler to release
+    * the resource in the event of an exception being raised during the computation,
+    * or in case of cancellation.
+    *
+    * In comparison with the simpler [[bracket]] version, this one allows the
+    * caller to differentiate between normal termination and cancellation.
+    *
+    * The `release` function receives as input:
+    *
+    *  - `Left(None)` in case of cancellation
+    *  - `Left(Some(error))` in case `use` terminated with an error
+    *  - `Right(b)` in case of success
+    *
+    * @param use is a function that evaluates the resource yielded by the source,
+    *        yielding a result that will get generated by this function on
+    *        evaluation
+    *
+    * @param release is a function that gets called after `use` terminates,
+    *        either normally or in error, or if it gets cancelled, receiving
+    *        as input the resource that needs that needs release, along with
+    *        the result of `use` (cancellation, error or successful result)
+    */
+  final def bracketE[B](use: A => Task[B])(release: (A, Either[Option[Throwable], B]) => Task[Unit]): Task[B] =
```

Comment:
Just to be 100% clear, `Either[Option[Throwable], B]` here means, that
`Left(None)` is cancelled, 
`Left(Some(ex))` is an error, and
`Right(b)` is success, correct? :)

Reply:
Yes. I'm also waiting on that ADT defined in `cats-effect`, for whenever it will happen, to introduce another overload.

Reply:
👍